GetDKAN/dkan
# Dynamic API docs generator with plugins

This completely replaces our existing OpenAPI spec generation with a more dynamic and modular one. A new plugin type, DkanApiDocs, is available, and documents can add their own sections to the spec. The spec is validated against the OpenAPI 3 spec on generation. The metastore API is more fully documented, and all endpoints have working examples. 

## QA steps
- Open the API page (/api) on the QA site.
- Try all the endpoints. Test out the examples and check that responses are accurate and examples work as expected
- Note whether the endpoints are documented in a way that is clear
- Visit a dataset page, ensure that the API docs for specific dataset still work as expected.
- Check /api/1.yml, for a YAML version of the spec.
- Check /api/1?authentication=false and confirm that authenticated endpoints and related components have been removed.

## Known issues
- The dataset-specific docs really need a refactor. They use the new way to generate the spec but then use a bunch of unnecessary logic on top of it left over from the old version to filter down to the two endpoints we actually use.
- Dredd tests have been disabled. Dredd just does not support enough of OpenAPI 3 to be practical. See #3536 for a possible way forward.

## Other notes
- Introduced a new metastore storage method: getRange(). This should be useful for paginating through datasets on sites with a lot of content, where getAll() will be impractical.
- Changed `JsonResponseTrait::getResponseFromException()` to provide more information including an arbitrary "data" property that tries to extract information from the caught exception object. Currently just supports ValidationException.
